pkg/controller/storage/capability/capability_controller.go

```diff
@@ -0,0 +1,317 @@
+		workQueue:                    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "StorageClasses"),
+		csiAddressGetter:             csiAddressGetter,
+	}
+	storageClassInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.enqueueStorageClass,
+		UpdateFunc: func(old, new interface{}) {
+			newStorageClass := new.(*v1strorage.StorageClass)
+			oldStorageClass := old.(*v1strorage.StorageClass)
+			if newStorageClass.ResourceVersion == oldStorageClass.ResourceVersion {
+				return
+			}
+			controller.enqueueStorageClass(newStorageClass)
+		},
+		DeleteFunc: controller.enqueueStorageClass,
+	})
+	snapshotClassInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.enqueueSnapshotClass,
+		UpdateFunc: func(old, new interface{}) {
+			return
+		},
+		DeleteFunc: controller.enqueueSnapshotClass,
+	})
+	return controller
+}
+
+func (c *StorageCapabilityController) Start(stopCh <-chan struct{}) error {
+	return c.Run(5, stopCh)
+}
+
+func (c *StorageCapabilityController) Run(threadCnt int, stopCh <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer c.workQueue.ShutDown()
+
+	isValid, err := c.IsValidKubernetesVersion()
+	if err != nil {
+		return err
+	}
+	if !isValid {
+		klog.V(0).Infof("StorageCapability controller minimal kubernetes version %s, skipped.", MinimalKubernetesVersion)
+		return nil
+	}
+	// Wait for the caches to be synced before starting workers
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.storageClassSynced, c.snapshotClassSynced, c.storageClassCapabilitySynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	for i := 0; i < threadCnt; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+	klog.Info("Started workers")
+	<-stopCh
+	klog.Info("Shutting down workers")
+	return nil
+}
+
+func (c *StorageCapabilityController) enqueueStorageClass(obj interface{}) {
+	storageClass := obj.(*v1strorage.StorageClass)
+	if !fileExist(c.csiAddressGetter(storageClass.Provisioner)) {
+		klog.V(4).Infof("CSI address of storage class: %s, provisioner :%s not exist", storageClass.Name, storageClass.Provisioner)
+		return
+	}
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	c.workQueue.Add(key)
+}
+
+func (c *StorageCapabilityController) enqueueSnapshotClass(obj interface{}) {
+	snapshotClass := obj.(*snapapi.VolumeSnapshotClass)
+	if !fileExist(c.csiAddressGetter(snapshotClass.Driver)) {
+		klog.V(4).Infof("CSI address of snapshot class: %s, driver:%s not exist", snapshotClass.Name, snapshotClass.Driver)
+		return
+	}
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	c.workQueue.Add(key)
```

Comment:
enqueued storageClass key in function `enqueueStorageClass`, and here enqueued snapshotClass key, there are two different types of object,  why put them in one queue?

Reply:
If the snaposhtClass with the same name with the storageClass was created or deleted, I will handler the StorageClassCapability with same name. The handler logic is the same, so i put  them in one queue.